fastscanner/services/indicators/service.py

```diff
@@ -136,18 +157,95 @@ async def handle(self, channel_id: str, data: dict[Any, Any]) -> None:
             if "timestamp" not in data:
                 logger.warning(f"Missing timestamp in message from {channel_id}")
                 return
+
             ts = pd.to_datetime(int(data["timestamp"]), unit="ms", utc=True).tz_convert(
                 LOCAL_TIMEZONE_STR
             )
             new_row = pd.Series(data, name=ts)
 
-            for indicator in self._indicators:
-                new_row = await indicator.extend_realtime(self._symbol, new_row)
-
-            self._handler.handle(self._symbol, new_row)
+            await self._add_to_buffer(new_row)
 
         except Exception as e:
             logger.error(
                 f"[Handler Error] Failed processing message from {channel_id}: {e}"
             )
+
+    async def _add_to_buffer(self, new_row: pd.Series):
+        assert isinstance(new_row.name, pd.Timestamp)
+        ts: pd.Timestamp = new_row.name
+        candle_start = self._get_interval_start(ts)
+        logger.debug(
+            f"[{self._symbol}] Buffer updated at {new_row.name} | Buffer keys: {[k for k in self._buffer.keys()]}"
+        )
+
+        if (
+            not self._buffer
+            or self._get_interval_start(next(iter(self._buffer))) != candle_start
+        ):
+            self._buffer.clear()
+
+            if self._timeout_task is not None and not self._timeout_task.done():
+                self._timeout_task.cancel()
+                try:
+                    await self._timeout_task
+                except asyncio.CancelledError:
+                    pass
+
+        self._buffer[ts] = new_row
+
+        if len(self._buffer) == self._expected_count:
+            await self._flush()
+        elif self._timeout_task is None or self._timeout_task.done():
+            self._timeout_task = asyncio.create_task(
+                self._flush_after_timeout(candle_start)
+            )
+
+    async def _flush_after_timeout(self, candle_start: pd.Timestamp):
+        now = ClockRegistry.clock.now()
+        candle_end = candle_start + pd.Timedelta(self._freq)
+        timeout_time = candle_end + pd.Timedelta(seconds=self._candle_timeout)
+        sleep_duration = (timeout_time - now).total_seconds()
+
+        if sleep_duration > 0:
+            await asyncio.sleep(sleep_duration)
+
+        await self._flush()
+
+    async def _flush(self):
+        if not self._buffer:
+            return
+
+        df = pd.DataFrame(self._buffer.values())
+        if df.empty:
             return
+
+        candle_start = self._get_interval_start(df.index[0])
+        logger.info(
+            f"[{self._symbol}] Flushing {self._freq} candle | Interval start: {candle_start} | Buffer size: {len(self._buffer)}"
+        )
+        resample_map = {
+            col: (
+                (lambda s: s.iloc[0])
+                if agg == "first"
+                else (lambda s: s.iloc[-1]) if agg == "last" else agg
+            )
+            for col, agg in CandleCol.RESAMPLE_MAP.items()
+        }
```

Comment:
This condition will be very confusing for a future developer. It's fine to leave it as you had implemented it before. 
Something like
```
pd.Series({
     CandleCol.OPEN: ...
     CandleCol.CLOSE: ...
```